### ISSUE TITLE
Build is failing due to jcenter dependency disappearing

### ISSUE THREAD

**Issue by Nick-Lucas** [react-native-community/react-native-share#386]
Our build has been failing since yesterday and pointing the finger at react-native-share
```
* What went wrong:
A problem occurred configuring project ':react-native-share'.
> Could not resolve all files for configuration ':react-native-share:classpath'.
   > Could not find lint-gradle-api.jar (com.android.tools.lint:lint-gradle-api:26.1.2).
     Searched in the following locations:
         https://jcenter.bintray.com/com/android/tools/lint/lint-gradle-api/26.1.2/lint-gradle-api-26.1.2.jar
```

It looks like this dependency has disappeared from bintray, though the .pom file is still in the folder
https://jcenter.bintray.com/com/android/tools/lint/lint-gradle-api/26.1.2/

https://jcenter.bintray.com/com/android/tools/lint/lint-gradle-api/26.1.2/lint-gradle-api-26.1.2.jar
```
{
  "errors" : [ {
    "status" : 404,
    "message" : "Could not find resource"
  } ]
}
```

I realise this is likely to be an upstream issue, but am hoping that someone here has an idea what we can do to fix this, since it's coming through react-native-share?

### Environment
- **React Native version**: 0.53.3
- **React Native platform + platform version**: Android build
- **Gradle version**:  4.4

### react-native-share
**Version**: 1.1.2

**Comment by AdriaRios** [comment]
+1

Adding `google()` on `build.graddle` fix the issue:
```
buildscript {
    repositories {
        google()
        jcenter()
        maven {
            url 'https://maven.google.com/'
            name 'Google'
        }
         maven { url 'https://maven.fabric.io/public' }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.1.3'
        classpath 'com.google.gms:google-services:4.0.1'
        classpath 'io.fabric.tools:gradle:1.25.4'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}
```

Also I have notice that the code on `master` is different, maybe there are a pending release?

**Comment by Nick-Lucas** [comment]
Hmmm that doesn't seem to have helped us.

our `allprojects` config was already configured that way, and I also tried adding to the `buildscript` config but the issue persisted.

Here's our full root build.gradle

```
buildscript {
    repositories {
        maven {
            url "https://maven.google.com/"
        }

        // WE TRIED ADDING THIS LINE
        google()

        jcenter()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.1.2'
        classpath 'com.google.gms:google-services:3.2.1'
    }
}

allprojects {
    repositories {
        mavenLocal()
        google()
        jcenter()
        maven {
            url "$rootDir/../node_modules/react-native/android"
        }
        maven {
            url "https://maven.google.com"
        }
    }
}
```

**Comment by AdriaRios** [comment]
I mean the `build.graddle` of `react-native-share`:

`node_modules/react-native-share/android/build.graddle`

**Comment by Nick-Lucas** [comment]
Oh interesting! I'll do some investigation and see if we can bodge this up for now

**Comment by mikehardy** [comment]
If bintray isn't carrying the android support libraries for some reason, there's nothing gradle (the android build/package tool) can do if you don't provide it a way to resolve the dependency somehow. It's as if npmjs.org decided to delete react-native-share, what can we do? Point it to somewhere that has it.

Who has this dependency guaranteed? The google repo. So somewhere the google repo isn't in the resolution path, and must be added (preferably first - they're more reliable than most...)

Adding the google repository is either via 'google()' in modern gradle versions or the 
```
maven {
            url "https://maven.google.com/"
        }
```
...stanza in older gradles (you don't need both - it's just that it didn't get a first-class name of `google()` until gradle 4.1) is going to be what resolves it but you'll need to add it in the "main project"

Then the question is where. Looks to me like this module has a valid pointer to google's maven repo in 1.1.2:

https://github.com/mikehardy/react-native-share/blob/83a3ec5e7bbc77f6681f0a96779692ba9a0e283e/android/build.gradle

If you want to check to see if the code in master will work, regardless, you can alter the react-native-share dependency as a test to `"git+https://github.com/react-native-community/react-native-share.git"` and run an `npm install` - that's how you ask npm (or yarn) to pull github master of any dependency, as opposed to using release packages.

I don't see a difference between the current release and master though so I think this dependency problem might be elsewhere?

**Comment by Nick-Lucas** [comment]
Confirmed as above, putting react-native-share's google() line above jcenter() fixes the issue. For now we are copying over a modified build.gradle into node-modules are part of the build process.

This isn't a great long term solution but should tide us over until a more robust release from this library can be released

**Comment by mikehardy** [comment]
Interesting - so if gradle (or maybe just the specific version of gradle used) gets a POM from a repo but the artifact 404s, it treats that as a failure meaning dependency order is very important. The official docs recommend google() at the top and that's what I've always done so this hasn't bitten me. I was the last to touch this file though, @jgcmarins - would you like a PR that flips the jcenter()/google() dependency?

I'm unable to find anything authoritative from the two parties (google or jcenter) on what order they recommend, but the gradle doc is explicit that order matters, so I think we get to make our own decision on who is more reliable with this issue seeming to be good evidence google() should go first...

Master still has jcenter() first:
https://github.com/mikehardy/react-native-share/blob/master/android/build.gradle

**Comment by mikehardy** [comment]
I've made a PR available in case it's wanted. Note if using modern gradle (3.2+) with  less-than-bleeding-edge React Native (<0.57) you will be missing a compatibility addition where react-native copies release build bundles into the new location gradle expects it: https://github.com/facebook/react-native/commit/a4fed6e203efd481175e61b7d2aa1c2b037a4279

In my own project I am not using RN57 yet, so I use their build-step workaround where for android release builds I pre-build the bundle into android/assets like so in the scripts area of package.json:

```
    "android-bundle": "node node_modules/react-native/local-cli/cli.js bundle --platform android --dev false --entry-file index.android.js --bundle-output android/kscore/src/main/assets/index.android.bundle --assets-dest android/kscore/src/main/res/",
```

**Comment by AdriaRios** [comment]
Thank you very much @mikehardy 

Also I have noticed that with the current order (with `jcenter()` first) and upgrading the graddle version it works:
`com.android.tools.build:gradle:3.1.3`

So we can conclude that is an issue related to version 3.1.2 and jcenter() dependency

**Comment by jgcmarins** [comment]
hey guys, thanks for this feedback
https://github.com/react-native-community/react-native-share/pull/387 would fix?
we can merge it into master, then anyone can test it?
if it's ok, we can have a new release

**Comment by AdriaRios** [comment]
Yes @jgcmarins 

I can confirm that the branch is working

**Comment by jgcmarins** [comment]
thanks everyone
let's test this on master branch and release a new version during the day

**Comment by jgcmarins** [comment]
done:
https://github.com/react-native-community/react-native-share/releases/tag/v1.1.3
https://www.npmjs.com/package/react-native-share

**Comment by anshumanburman** [comment]
@AdriaRios thanks

**Comment by mustafakucuk** [comment]
If it's give "Could not find method google() for arguments [] on repository container" error, replace google() to maven { url "https://maven.google.com" } in react-native-share/android/build.gradle

**Comment by mikehardy** [comment]
> If it's give "Could not find method google() for arguments [] on repository container" error, replace google() to maven { url "[maven.google.com](https://maven.google.com)" } in react-native-share/android/build.gradle

Yes - this will happen if using gradle version lower than 4.1 per the docs:

https://developer.android.com/studio/build/dependencies#google-maven

Gradle 4.1 is about 1.5years old now though, they just released 5.0 which I haven't tried yet, but 4.10.2 is nice - I'd move up
